### 1. Importing the necessary modules and accessing data from google drive

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import plotly.express as px
from datetime import datetime
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
import plotly.graph_objects as go

In [2]:
sales_path = "sales.csv"
stock_sensor_path = "sensor_stock_levels.csv"
storage_sensor_path = "sensor_storage_temperature.csv"

### 2. Exploring and cleaning the dataset using inbuilt-functions of the Pandas Module.

In [3]:
sales_df = pd.read_csv(sales_path)

In [4]:
sales_df.drop(columns=["Unnamed: 0"], inplace=True, errors='ignore')
sales_df.head(10)

transaction_id            timestamp  \
0  a1c82654-c52c-45b3-8ce8-4c2a1efe63ed  2022-03-02 09:51:38   
1  931ad550-09e8-4da6-beaa-8c9d17be9c60  2022-03-06 10:33:59   
2  ae133534-6f61-4cd6-b6b8-d1c1d8d90aea  2022-03-04 17:20:21   
3  157cebd9-aaf0-475d-8a11-7c8e0f5b76e4  2022-03-02 17:23:58   
4  a81a6cd3-5e0c-44a2-826c-aea43e46c514  2022-03-05 14:32:43   
5  b5b3c8b9-f496-484d-aa30-4f2efb5ed56c  2022-03-07 17:59:47   
6  4997b1ae-f5aa-4b9f-8fc8-22ad8f19837c  2022-03-07 19:36:57   
7  bfffee68-0736-42af-bd3e-4ca77541b0d6  2022-03-07 19:03:20   
8  ce50e984-90cd-4b47-8537-3a4e32e6d314  2022-03-07 11:34:32   
9  f0700cc9-e6f5-4b9a-add4-05e64b4f495b  2022-03-07 09:20:12   

                             product_id category customer_type  unit_price  \
0  3bc6c1ea-0198-46de-9ffd-514ae3338713    fruit          gold        3.99   
1  ad81b46c-bf38-41cf-9b54-5fe7f5eba93e    fruit      standard        3.99   
2  7c55cbd4-f306-4c04-a030-628cbe7867c1    fruit       premium        0.19   
3  80da8348-1707-403f-8be7-9e6deeccc883    fruit          gold        0.19   
4  7f5e86e6-f06f-45f6-bf44-27b095c9ad1d    fruit         basic        4.49   
5  3bc6c1ea-0198-46de-9ffd-514ae3338713    fruit      standard        3.99   
6  14736243-d346-438f-9535-d80fcb9f3882    fruit      standard        1.49   
7  0ddc2379-adba-4fb0-aa97-19fcafc738a1    fruit         basic        3.99   
8  ad81b46c-bf38-41cf-9b54-5fe7f5eba93e    fruit    non-member        3.99   
9  35e00193-aa27-4121-87f4-6ac0d8a8671f    fruit       premium        0.49   

   quantity  total payment_type  
0         2   7.98     e-wallet  
1         1   3.99     e-wallet  
2         2   0.38     e-wallet  
3         4   0.76     e-wallet  
4         2   8.98   debit card  
5         4  15.96         cash  
6         4   5.96     e-wallet  
7         4  15.96  credit card  
8         1   3.99  credit card  
9         3   1.47     e-wallet

In [5]:
stock_df = pd.read_csv(stock_sensor_path)
stock_df.drop(columns=["Unnamed: 0"], inplace=True, errors='ignore')
stock_df.head(10)

id            timestamp  \
0  4220e505-c247-478d-9831-6b9f87a4488a  2022-03-07 12:13:02   
1  f2612b26-fc82-49ea-8940-0751fdd4d9ef  2022-03-07 16:39:46   
2  989a287f-67e6-4478-aa49-c3a35dac0e2e  2022-03-01 18:17:43   
3  af8e5683-d247-46ac-9909-1a77bdebefb2  2022-03-02 14:29:09   
4  08a32247-3f44-4002-85fb-c198434dd4bb  2022-03-02 13:46:18   
5  98550364-b9c1-42e2-8852-a0e884daf960  2022-03-01 14:51:16   
6  35da168d-715e-4ca1-8755-d317cee4e9b9  2022-03-03 12:59:14   
7  b00838b1-59a6-4f85-81fe-986ed0e46fc2  2022-03-01 12:35:30   
8  4478f269-e631-465f-b490-0d8629dd138d  2022-03-07 13:45:06   
9  ca95cb50-26cb-4d87-8819-baefd68de5d9  2022-03-05 09:18:11   

                             product_id  estimated_stock_pct  
0  f658605e-75f3-4fed-a655-c0903f344427                 0.75  
1  de06083a-f5c0-451d-b2f4-9ab88b52609d                 0.48  
2  ce8f3a04-d1a4-43b1-a7c2-fa1b8e7674c8                 0.58  
3  c21e3ba9-92a3-4745-92c2-6faef73223f7                 0.79  
4  7f478817-aa5b-44e9-9059-8045228c9eb0                 0.22  
5  0f493675-d59d-45dd-9406-084932bf41c1                 0.04  
6  e22ea344-cea3-42fe-bb97-9a99fa7c6987                 0.03  
7  a3ec02dd-a6be-4fc0-a8d1-ef34a2addca0                 0.56  
8  21f8aa00-bc6e-4c9d-8acf-2d3bb740f30b                 0.89  
9  de06083a-f5c0-451d-b2f4-9ab88b52609d                 0.99

In [6]:
storage_df = pd.read_csv(storage_sensor_path)
storage_df.drop(columns=["Unnamed: 0"], inplace=True, errors='ignore')
storage_df.head(10)

id            timestamp  temperature
0  d1ca1ef8-0eac-42fc-af80-97106efc7b13  2022-03-07 15:55:20         2.96
1  4b8a66c4-0f3a-4f16-826f-8cf9397e9d18  2022-03-01 09:18:22         1.88
2  3d47a0c7-1e72-4512-812f-b6b5d8428cf3  2022-03-04 15:12:26         1.78
3  9500357b-ce15-424a-837a-7677b386f471  2022-03-02 12:30:42         2.18
4  c4b61fec-99c2-4c6d-8e5d-4edd8c9632fa  2022-03-05 09:09:33         1.38
5  125583af-e280-4627-9cc6-62413af94c6b  2022-03-06 09:39:53         0.04
6  3ef1f91d-f7ef-4bf2-a0f7-2c3720a7432c  2022-03-05 19:21:53       -28.34
7  5131c3ed-4f6b-4f52-93c9-fa2717499a80  2022-03-03 19:09:48        -2.47
8  0fd6edd9-36d3-4c7c-8184-53e746365067  2022-03-05 15:45:46         2.48
9  04766f4f-a431-4853-92c9-a6b28b7f7abb  2022-03-04 15:18:25        -3.03

In [7]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7829 entries, 0 to 7828
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   transaction_id  7829 non-null   object 
 1   timestamp       7829 non-null   object 
 2   product_id      7829 non-null   object 
 3   category        7829 non-null   object 
 4   customer_type   7829 non-null   object 
 5   unit_price      7829 non-null   float64
 6   quantity        7829 non-null   int64  
 7   total           7829 non-null   float64
 8   payment_type    7829 non-null   object 
dtypes: float64(2), int64(1), object(6)
memory usage: 550.6+ KB


In [8]:
stock_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   15000 non-null  object 
 1   timestamp            15000 non-null  object 
 2   product_id           15000 non-null  object 
 3   estimated_stock_pct  15000 non-null  float64
dtypes: float64(1), object(3)
memory usage: 468.9+ KB


In [9]:
storage_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23890 entries, 0 to 23889
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           23890 non-null  object 
 1   timestamp    23890 non-null  object 
 2   temperature  23890 non-null  float64
dtypes: float64(1), object(2)
memory usage: 560.0+ KB


### 3. Coverting 'timestamp' column from object type to datetime format

In [10]:
def convert_to_datetime(data: pd.DataFrame = None, column: str = None):
  dummy = data.copy()
  dummy[column] = pd.to_datetime(dummy[column], format='%Y-%m-%d %H:%M:%S')
  return dummy

In [11]:
sales_df = convert_to_datetime(sales_df, 'timestamp')
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7829 entries, 0 to 7828
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   transaction_id  7829 non-null   object        
 1   timestamp       7829 non-null   datetime64[ns]
 2   product_id      7829 non-null   object        
 3   category        7829 non-null   object        
 4   customer_type   7829 non-null   object        
 5   unit_price      7829 non-null   float64       
 6   quantity        7829 non-null   int64         
 7   total           7829 non-null   float64       
 8   payment_type    7829 non-null   object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(5)
memory usage: 550.6+ KB


In [12]:
stock_df = convert_to_datetime(stock_df, 'timestamp')
stock_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   id                   15000 non-null  object        
 1   timestamp            15000 non-null  datetime64[ns]
 2   product_id           15000 non-null  object        
 3   estimated_stock_pct  15000 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(2)
memory usage: 468.9+ KB


In [13]:
storage_df = convert_to_datetime(storage_df, 'timestamp')
storage_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23890 entries, 0 to 23889
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id           23890 non-null  object        
 1   timestamp    23890 non-null  datetime64[ns]
 2   temperature  23890 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 560.0+ KB


In [14]:
def convert_timestamp_to_hourly(data: pd.DataFrame = None, column: str = None):
  dummy = data.copy()
  new_ts = dummy[column].tolist()
  new_ts = [i.strftime('%Y-%m-%d %H:00:00') for i in new_ts]
  new_ts = [datetime.strptime(i, '%Y-%m-%d %H:00:00') for i in new_ts]
  dummy[column] = new_ts
  return dummy

In [15]:
sales_df = convert_timestamp_to_hourly(sales_df, 'timestamp')
sales_df.head()

transaction_id           timestamp  \
0  a1c82654-c52c-45b3-8ce8-4c2a1efe63ed 2022-03-02 09:00:00   
1  931ad550-09e8-4da6-beaa-8c9d17be9c60 2022-03-06 10:00:00   
2  ae133534-6f61-4cd6-b6b8-d1c1d8d90aea 2022-03-04 17:00:00   
3  157cebd9-aaf0-475d-8a11-7c8e0f5b76e4 2022-03-02 17:00:00   
4  a81a6cd3-5e0c-44a2-826c-aea43e46c514 2022-03-05 14:00:00   

                             product_id category customer_type  unit_price  \
0  3bc6c1ea-0198-46de-9ffd-514ae3338713    fruit          gold        3.99   
1  ad81b46c-bf38-41cf-9b54-5fe7f5eba93e    fruit      standard        3.99   
2  7c55cbd4-f306-4c04-a030-628cbe7867c1    fruit       premium        0.19   
3  80da8348-1707-403f-8be7-9e6deeccc883    fruit          gold        0.19   
4  7f5e86e6-f06f-45f6-bf44-27b095c9ad1d    fruit         basic        4.49   

   quantity  total payment_type  
0         2   7.98     e-wallet  
1         1   3.99     e-wallet  
2         2   0.38     e-wallet  
3         4   0.76     e-wallet  
4         2   8.98   debit card

In [16]:
stock_df = convert_timestamp_to_hourly(stock_df, 'timestamp')
stock_df.head()

id           timestamp  \
0  4220e505-c247-478d-9831-6b9f87a4488a 2022-03-07 12:00:00   
1  f2612b26-fc82-49ea-8940-0751fdd4d9ef 2022-03-07 16:00:00   
2  989a287f-67e6-4478-aa49-c3a35dac0e2e 2022-03-01 18:00:00   
3  af8e5683-d247-46ac-9909-1a77bdebefb2 2022-03-02 14:00:00   
4  08a32247-3f44-4002-85fb-c198434dd4bb 2022-03-02 13:00:00   

                             product_id  estimated_stock_pct  
0  f658605e-75f3-4fed-a655-c0903f344427                 0.75  
1  de06083a-f5c0-451d-b2f4-9ab88b52609d                 0.48  
2  ce8f3a04-d1a4-43b1-a7c2-fa1b8e7674c8                 0.58  
3  c21e3ba9-92a3-4745-92c2-6faef73223f7                 0.79  
4  7f478817-aa5b-44e9-9059-8045228c9eb0                 0.22

In [17]:
storage_df = convert_timestamp_to_hourly(storage_df, 'timestamp')
storage_df.head()

id           timestamp  temperature
0  d1ca1ef8-0eac-42fc-af80-97106efc7b13 2022-03-07 15:00:00         2.96
1  4b8a66c4-0f3a-4f16-826f-8cf9397e9d18 2022-03-01 09:00:00         1.88
2  3d47a0c7-1e72-4512-812f-b6b5d8428cf3 2022-03-04 15:00:00         1.78
3  9500357b-ce15-424a-837a-7677b386f471 2022-03-02 12:00:00         2.18
4  c4b61fec-99c2-4c6d-8e5d-4edd8c9632fa 2022-03-05 09:00:00         1.38

### 4.Preparing data for Training and Testing

In [18]:
sales_agg = sales_df.groupby(['timestamp', 'product_id']).agg({'quantity': 'sum'}).reset_index()
sales_agg.head()

timestamp                            product_id  quantity
0 2022-03-01 09:00:00  00e120bb-89d6-4df5-bc48-a051148e3d03         3
1 2022-03-01 09:00:00  01f3cdd9-8e9e-4dff-9b5c-69698a0388d0         3
2 2022-03-01 09:00:00  03a2557a-aa12-4add-a6d4-77dc36342067         3
3 2022-03-01 09:00:00  049b2171-0eeb-4a3e-bf98-0c290c7821da         7
4 2022-03-01 09:00:00  04da844d-8dba-4470-9119-e534d52a03a0        11

In [19]:
stock_agg = stock_df.groupby(['timestamp', 'product_id']).agg({'estimated_stock_pct': 'mean'}).reset_index()
stock_agg.head()

timestamp                            product_id  \
0 2022-03-01 09:00:00  00e120bb-89d6-4df5-bc48-a051148e3d03   
1 2022-03-01 09:00:00  01f3cdd9-8e9e-4dff-9b5c-69698a0388d0   
2 2022-03-01 09:00:00  01ff0803-ae73-4234-971d-5713c97b7f4b   
3 2022-03-01 09:00:00  0363eb21-8c74-47e1-a216-c37e565e5ceb   
4 2022-03-01 09:00:00  03f0b20e-3b5b-444f-bc39-cdfa2523d4bc   

   estimated_stock_pct  
0                 0.89  
1                 0.14  
2                 0.67  
3                 0.82  
4                 0.05

In [20]:
storage_agg = storage_df.groupby(['timestamp']).agg({'temperature': 'mean'}).reset_index()
storage_agg.head()

timestamp  temperature
0 2022-03-01 09:00:00    -0.028850
1 2022-03-01 10:00:00     1.284314
2 2022-03-01 11:00:00    -0.560000
3 2022-03-01 12:00:00    -0.537721
4 2022-03-01 13:00:00    -0.188734

In [21]:
merged_df = stock_agg.merge(sales_agg, on=['timestamp', 'product_id'], how='left')
merged_df.head()

timestamp                            product_id  \
0 2022-03-01 09:00:00  00e120bb-89d6-4df5-bc48-a051148e3d03   
1 2022-03-01 09:00:00  01f3cdd9-8e9e-4dff-9b5c-69698a0388d0   
2 2022-03-01 09:00:00  01ff0803-ae73-4234-971d-5713c97b7f4b   
3 2022-03-01 09:00:00  0363eb21-8c74-47e1-a216-c37e565e5ceb   
4 2022-03-01 09:00:00  03f0b20e-3b5b-444f-bc39-cdfa2523d4bc   

   estimated_stock_pct  quantity  
0                 0.89       3.0  
1                 0.14       3.0  
2                 0.67       NaN  
3                 0.82       NaN  
4                 0.05       NaN

In [22]:
merged_df = merged_df.merge(storage_agg, on='timestamp', how='left')
merged_df.head()

timestamp                            product_id  \
0 2022-03-01 09:00:00  00e120bb-89d6-4df5-bc48-a051148e3d03   
1 2022-03-01 09:00:00  01f3cdd9-8e9e-4dff-9b5c-69698a0388d0   
2 2022-03-01 09:00:00  01ff0803-ae73-4234-971d-5713c97b7f4b   
3 2022-03-01 09:00:00  0363eb21-8c74-47e1-a216-c37e565e5ceb   
4 2022-03-01 09:00:00  03f0b20e-3b5b-444f-bc39-cdfa2523d4bc   

   estimated_stock_pct  quantity  temperature  
0                 0.89       3.0     -0.02885  
1                 0.14       3.0     -0.02885  
2                 0.67       NaN     -0.02885  
3                 0.82       NaN     -0.02885  
4                 0.05       NaN     -0.02885

In [23]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10845 entries, 0 to 10844
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   timestamp            10845 non-null  datetime64[ns]
 1   product_id           10845 non-null  object        
 2   estimated_stock_pct  10845 non-null  float64       
 3   quantity             3067 non-null   float64       
 4   temperature          10845 non-null  float64       
dtypes: datetime64[ns](1), float64(3), object(1)
memory usage: 423.8+ KB


In [24]:
merged_df['quantity'] = merged_df['quantity'].fillna(0)
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10845 entries, 0 to 10844
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   timestamp            10845 non-null  datetime64[ns]
 1   product_id           10845 non-null  object        
 2   estimated_stock_pct  10845 non-null  float64       
 3   quantity             10845 non-null  float64       
 4   temperature          10845 non-null  float64       
dtypes: datetime64[ns](1), float64(3), object(1)
memory usage: 423.8+ KB


In [25]:
product_categories = sales_df[['product_id', 'category']]
product_categories = product_categories.drop_duplicates()
product_price = sales_df[['product_id', 'unit_price']]
product_price = product_price.drop_duplicates()

In [26]:
merged_df = merged_df.merge(product_categories, on="product_id", how="left")
merged_df.head()

timestamp                            product_id  \
0 2022-03-01 09:00:00  00e120bb-89d6-4df5-bc48-a051148e3d03   
1 2022-03-01 09:00:00  01f3cdd9-8e9e-4dff-9b5c-69698a0388d0   
2 2022-03-01 09:00:00  01ff0803-ae73-4234-971d-5713c97b7f4b   
3 2022-03-01 09:00:00  0363eb21-8c74-47e1-a216-c37e565e5ceb   
4 2022-03-01 09:00:00  03f0b20e-3b5b-444f-bc39-cdfa2523d4bc   

   estimated_stock_pct  quantity  temperature       category  
0                 0.89       3.0     -0.02885        kitchen  
1                 0.14       3.0     -0.02885     vegetables  
2                 0.67       0.0     -0.02885  baby products  
3                 0.82       0.0     -0.02885      beverages  
4                 0.05       0.0     -0.02885           pets

In [27]:
merged_df = merged_df.merge(product_price, on="product_id", how="left")
merged_df.head()

timestamp                            product_id  \
0 2022-03-01 09:00:00  00e120bb-89d6-4df5-bc48-a051148e3d03   
1 2022-03-01 09:00:00  01f3cdd9-8e9e-4dff-9b5c-69698a0388d0   
2 2022-03-01 09:00:00  01ff0803-ae73-4234-971d-5713c97b7f4b   
3 2022-03-01 09:00:00  0363eb21-8c74-47e1-a216-c37e565e5ceb   
4 2022-03-01 09:00:00  03f0b20e-3b5b-444f-bc39-cdfa2523d4bc   

   estimated_stock_pct  quantity  temperature       category  unit_price  
0                 0.89       3.0     -0.02885        kitchen       11.19  
1                 0.14       3.0     -0.02885     vegetables        1.49  
2                 0.67       0.0     -0.02885  baby products       14.19  
3                 0.82       0.0     -0.02885      beverages       20.19  
4                 0.05       0.0     -0.02885           pets        8.19

In [28]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10845 entries, 0 to 10844
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   timestamp            10845 non-null  datetime64[ns]
 1   product_id           10845 non-null  object        
 2   estimated_stock_pct  10845 non-null  float64       
 3   quantity             10845 non-null  float64       
 4   temperature          10845 non-null  float64       
 5   category             10845 non-null  object        
 6   unit_price           10845 non-null  float64       
dtypes: datetime64[ns](1), float64(4), object(2)
memory usage: 593.2+ KB


In [29]:
merged_df['timestamp_day_of_month'] = merged_df['timestamp'].dt.day
merged_df['timestamp_day_of_week'] = merged_df['timestamp'].dt.dayofweek
merged_df['timestamp_hour'] = merged_df['timestamp'].dt.hour
merged_df.drop(columns=['timestamp'], inplace=True)
merged_df.head()

product_id  estimated_stock_pct  quantity  \
0  00e120bb-89d6-4df5-bc48-a051148e3d03                 0.89       3.0   
1  01f3cdd9-8e9e-4dff-9b5c-69698a0388d0                 0.14       3.0   
2  01ff0803-ae73-4234-971d-5713c97b7f4b                 0.67       0.0   
3  0363eb21-8c74-47e1-a216-c37e565e5ceb                 0.82       0.0   
4  03f0b20e-3b5b-444f-bc39-cdfa2523d4bc                 0.05       0.0   

   temperature       category  unit_price  timestamp_day_of_month  \
0     -0.02885        kitchen       11.19                       1   
1     -0.02885     vegetables        1.49                       1   
2     -0.02885  baby products       14.19                       1   
3     -0.02885      beverages       20.19                       1   
4     -0.02885           pets        8.19                       1   

   timestamp_day_of_week  timestamp_hour  
0                      1               9  
1                      1               9  
2                      1               9  
3                      1               9  
4                      1               9

In [30]:
merged_df = pd.get_dummies(merged_df, columns=['category'])
merged_df.head()

product_id  estimated_stock_pct  quantity  \
0  00e120bb-89d6-4df5-bc48-a051148e3d03                 0.89       3.0   
1  01f3cdd9-8e9e-4dff-9b5c-69698a0388d0                 0.14       3.0   
2  01ff0803-ae73-4234-971d-5713c97b7f4b                 0.67       0.0   
3  0363eb21-8c74-47e1-a216-c37e565e5ceb                 0.82       0.0   
4  03f0b20e-3b5b-444f-bc39-cdfa2523d4bc                 0.05       0.0   

   temperature  unit_price  timestamp_day_of_month  timestamp_day_of_week  \
0     -0.02885       11.19                       1                      1   
1     -0.02885        1.49                       1                      1   
2     -0.02885       14.19                       1                      1   
3     -0.02885       20.19                       1                      1   
4     -0.02885        8.19                       1                      1   

   timestamp_hour  category_baby products  category_baked goods  ...  \
0               9                   False                 False  ...   
1               9                   False                 False  ...   
2               9                    True                 False  ...   
3               9                   False                 False  ...   
4               9                   False                 False  ...   

   category_meat  category_medicine  category_packaged foods  \
0          False              False                    False   
1          False              False                    False   
2          False              False                    False   
3          False              False                    False   
4          False              False                    False   

   category_personal care  category_pets  category_refrigerated items  \
0                   False          False                        False   
1                   False          False                        False   
2                   False          False                        False   
3                   False          False                        False   
4                   False           True                        False   

   category_seafood  category_snacks  category_spices and herbs  \
0             False            False                      False   
1             False            False                      False   
2             False            False                      False   
3             False            False                      False   
4             False            False                      False   

   category_vegetables  
0                False  
1                 True  
2                False  
3                False  
4                False  

[5 rows x 30 columns]

In [31]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10845 entries, 0 to 10844
Data columns (total 30 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   product_id                      10845 non-null  object 
 1   estimated_stock_pct             10845 non-null  float64
 2   quantity                        10845 non-null  float64
 3   temperature                     10845 non-null  float64
 4   unit_price                      10845 non-null  float64
 5   timestamp_day_of_month          10845 non-null  int32  
 6   timestamp_day_of_week           10845 non-null  int32  
 7   timestamp_hour                  10845 non-null  int32  
 8   category_baby products          10845 non-null  bool   
 9   category_baked goods            10845 non-null  bool   
 10  category_baking                 10845 non-null  bool   
 11  category_beverages              10845 non-null  bool   
 12  category_canned foods           

In [32]:
merged_df.drop(columns=['product_id'], inplace=True)
merged_df.head()

estimated_stock_pct  quantity  temperature  unit_price  \
0                 0.89       3.0     -0.02885       11.19   
1                 0.14       3.0     -0.02885        1.49   
2                 0.67       0.0     -0.02885       14.19   
3                 0.82       0.0     -0.02885       20.19   
4                 0.05       0.0     -0.02885        8.19   

   timestamp_day_of_month  timestamp_day_of_week  timestamp_hour  \
0                       1                      1               9   
1                       1                      1               9   
2                       1                      1               9   
3                       1                      1               9   
4                       1                      1               9   

   category_baby products  category_baked goods  category_baking  ...  \
0                   False                 False            False  ...   
1                   False                 False            False  ...   
2                    True                 False            False  ...   
3                   False                 False            False  ...   
4                   False                 False            False  ...   

   category_meat  category_medicine  category_packaged foods  \
0          False              False                    False   
1          False              False                    False   
2          False              False                    False   
3          False              False                    False   
4          False              False                    False   

   category_personal care  category_pets  category_refrigerated items  \
0                   False          False                        False   
1                   False          False                        False   
2                   False          False                        False   
3                   False          False                        False   
4                   False           True                        False   

   category_seafood  category_snacks  category_spices and herbs  \
0             False            False                      False   
1             False            False                      False   
2             False            False                      False   
3             False            False                      False   
4             False            False                      False   

   category_vegetables  
0                False  
1                 True  
2                False  
3                False  
4                False  

[5 rows x 29 columns]

In [33]:
merged_df.to_csv('Cleaned and combined dataset.csv', index=False)

In [34]:
x = merged_df.drop(columns=['estimated_stock_pct'])
y = merged_df['estimated_stock_pct']
print(x.shape)
print(y.shape)

(10845, 28)
(10845,)


### 5. Implementing a K-fold cross-validation using a RandomForestRegressor model.

#### Training and evaluating the model on different folds of the dataset, computing the Mean Absolute Error (MAE) as a measure of accuracy, and finally, calculating the average MAE across all folds.

#### The data is preprocessed by scaling using StandardScaler before training the model.

In [35]:
K = 10
split = 0.75
accuracy = []
# Instantiate algorithm
scaler = StandardScaler()

# Create training and test samples
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size=split, random_state=42)
# Scale X data, we scale the data because it helps the algorithm to converge
# and helps the algorithm to not be greedy with large values
scaler.fit(x_train)
x_train = scaler.transform(x_train)
x_test = scaler.transform(x_test)

In [36]:
model = RandomForestRegressor()
for fold in range(0, K):

  # Train model
  trained_model = model.fit(x_train, y_train)

  # Generate predictions on test sample
  y_pred = trained_model.predict(x_test)

  # Compute accuracy, using mean absolute error
  mae = mean_absolute_error(y_true=y_test, y_pred=y_pred)
  accuracy.append(mae)
  print(f"Fold {fold + 1}: MAE = {mae:.3f}")

print(f"Average MAE: {(sum(accuracy) / len(accuracy)):.2f}")

Fold 1: MAE = 0.237
Fold 2: MAE = 0.236
Fold 3: MAE = 0.236
Fold 4: MAE = 0.237
Fold 5: MAE = 0.236
Fold 6: MAE = 0.236
Fold 7: MAE = 0.236
Fold 8: MAE = 0.236
Fold 9: MAE = 0.236
Fold 10: MAE = 0.236
Average MAE: 0.24


### 6. Visualizing the feature importances of the above RandomForestRegressor model

In [40]:
features = [i.split("__")[0] for i in x.columns]
importances = model.feature_importances_
indices = np.argsort(importances)

fig = go.Figure()

fig.add_trace(go.Bar(
    y=[features[i] for i in indices],
    x=importances[indices],
    text=[f'{imp:.3f}' for imp in importances[indices]],
    textposition='auto',  # Display text inside bars
    orientation='h',
    marker=dict(color='lightgreen'),
))

fig.update_layout(
    title='Feature Importances identified by RandomForestRegressor Model',
    yaxis=dict(title='Features'),
    xaxis=dict(title='Relative Importance'),
    height=900,
    width=1300,
)

fig.show()

### 7. Implementing a K-fold cross-validation using a DecisionTreeRegressor model.

In [38]:
model = DecisionTreeRegressor()
for fold in range(0, K):

  # Train model
  trained_model = model.fit(x_train, y_train)

  # Generate predictions on test sample
  y_pred = trained_model.predict(x_test)

  # Compute accuracy, using mean absolute error
  mae = mean_absolute_error(y_true=y_test, y_pred=y_pred)
  accuracy.append(mae)
  print(f"Fold {fold + 1}: MAE = {mae:.3f}")

print(f"Average MAE: {(sum(accuracy) / len(accuracy)):.2f}")

Fold 1: MAE = 0.312
Fold 2: MAE = 0.311
Fold 3: MAE = 0.315
Fold 4: MAE = 0.312
Fold 5: MAE = 0.312
Fold 6: MAE = 0.312
Fold 7: MAE = 0.313
Fold 8: MAE = 0.313
Fold 9: MAE = 0.312
Fold 10: MAE = 0.313
Average MAE: 0.27


#### 8. Visualizing the feature importances of the above DecisionTreeRegressor model

In [41]:
features = [i.split("__")[0] for i in x.columns]
importances = model.feature_importances_
indices = np.argsort(importances)

fig = go.Figure()

fig.add_trace(go.Bar(
    y=[features[i] for i in indices],
    x=importances[indices],
    text=[f'{imp:.3f}' for imp in importances[indices]],
    textposition='auto',  # Display text inside bars
    orientation='h',
    marker=dict(color='lightgreen'),
))

fig.update_layout(
    title=f'Feature Importances - {model.__class__.__name__}',
    yaxis=dict(title='Features'),
    xaxis=dict(title='Relative Importance'),
    height=800,
    width=1000,
)

fig.show()

#### From above we can infer the following: -


*   Average MAE for RandomForestRegressor Model -> 0.24
*   Average MAE for DecisionTreeRegressor Model -> 0.27


#### The best model that gave prediction with least Average MAE is Linear Regression Model





# **From the plotly plots of Feature Importances of LinearRegressor Model and DecisionTreeRegressor Model we can infer that "“Yes, we can accurately predict the stock levels of products based on sales data and sensor data on an hourly basis in particular cases, in order to more intelligently procure products from our suppliers."**